Original/man/build-all.man
.\" ========================================================
.TH BUILD-ALL 1 "" "1.00"
.\" ========================================================
.SH NAME
build-all \(em build software distributions on remote systems in parallel
.\" ========================================================
.SH SYNOPSIS
.B build-all
[
.B \-\-?
]
[
.BI \-\-all " target(s)"
]
[
.BI \-\-check " target(s)"
]
.if n .ti +10n
[
.BI \-\-configure " options"
]
[
.B \-\-help
]
.if n .ti +10n
.if t .ti +\w'\fBbuild-all\fP\ 'u
[
.BI \-\-logdirectory " dir"
]
.if n .ti +10n
[
.BI \-\-on " [user@]host[:dir][,envfile] .\|.\|."
]
.if n .ti +10n
[
.BI \-\-source " dir(s)"
]
.if t .ti +\w'\fBbuild-all\fP\ 'u
[
.BI \-\-userhosts " file(s)"
]
.if n .ti +10n
[
.B \-\-version
]
package(s)
.\" ========================================================
.SH OPTIONS
.B build-all
options can be prefixed with either one or two hyphens, and
can be abbreviated to any unique prefix.  Thus,
.BR \-v ,
.BR \-ver ,
and
.B \-\-version
are equivalent.
.PP
To avoid confusion with options, if a filename begins with a
hyphen, it must be disguised by a leading absolute or
relative directory path, e.g.,
.I /tmp/-foo
or
.IR ./-foo .
.\" --------------------------------------------------------
.TP \w'\fB\-configure\fP\ \fIoptions\fP'u+3n
.BI \-\-all " target(s)"
Specify the default
.BR make (1)
targets to build the package [default: empty].
.\" --------------------------------------------------------
.TP
.B \-\-?
Same as
.BR \-\-help .
.\" --------------------------------------------------------
.TP
.BI \-\-check " target(s)"
Specify the default
.BR make (1)
targets to test the package [default:
.IR check ].
.\" --------------------------------------------------------
.TP
.BI \-\-configure " options"
Specify any options needed by the
.BR configure (1)
script.
.\" --------------------------------------------------------
.TP
.B \-\-help
Display a brief help message on
.IR stdout ,
giving a usage description, and then terminate immediately
with a success return code.
.\" --------------------------------------------------------
.TP
.BI \-\-logdirectory " dir"
Specify an alternate log file directory.  By default, logs
are created in a subdirectory named
.I logs/package-x.y.z
for package
.I package-x.y.z
in the directory in which the package archive is found.  If
the directory cannot be created or written to, a default log
directory is chosen and reported.
.\" --------------------------------------------------------
.TP
.BI \-\-on " [user@]host[:dir][,envfile] .\|.\|."
Specify an alternate list of hosts on which to build the
packages.  The list must be quoted if it contains spaces or
other shell metacharacters.  This option suppresses reading
the host list from the
.I $HOME/.build/userhosts
initialization file, allowing builds to be restricted to
a particular set of hosts.
.IP
In each host specification, the username may be omitted if
it is the same as on the current host.  The hostname is
mandatory.  The directory in which packages are unbundled
and built is given by
.IR dir :
a default directory is chosen and reported if that value is
omitted.  The optional environment file,
.IR envfile ,
on the initiating host
.RI ( not
the remote host), provides settings of environment variables
for
.BR configure (1),
in the form of
.I key=value
assignments.
.IP
This option may be given any number of times.
.\" --------------------------------------------------------
.TP
.BI \-\-source " dir(s)"
Specify an additional list of directories to search for
source archives.  The list must be quoted if it contains
spaces or other shell metacharacters.  That list will be
searched before the directories in the
.I $HOME/.build/directories
list, or the built-in list.
.IP
This option may be given any number of times.
.\" --------------------------------------------------------
.TP
.BI \-\-userhosts " file(s)"
Specify an alternate list of userhosts files to replace the
default initialization file,
.IR $HOME/.build/userhosts .
The list must be quoted if it contains spaces or other shell
metacharacters.
.IP
If a specified userhosts file is not readable as given, it
is looked for in
.IR $HOME/.build ,
.IP
This option may be given any number of times.
.IP
Settings from the
.B \-\-on
option override this option, so the two should be considered
mutually exclusive.
.IP
This option makes it convenient to create sets of build
hosts grouped by compiler, platform, or package, to
accommodate packages that are known to build only in certain
limited environments.
.\" --------------------------------------------------------
.TP
.B \-\-version
Display the program version number and release date on
.IR stdout ,
and then terminate immediately with a success return code.
.\" ========================================================
.SH DESCRIPTION
.B build-all
provides for automated software builds on one or more
hosts, in parallel, with build logs collected in
package-specific directories on the initiating host, using
the secure shell to communicate with the remote hosts.
.PP
The log directory is always reported on
.IR stdout ,
and in an email message.
.PP
Log files are named with the package name, version, build
host, and an ISO 8601:2000 odometer-style time stamp, like
this:
.IR package-x.y.z.hostname.yyyy.mm.dd.hh.mm.ss.log .
.PP
Packages should conform to GNU conventions for
software packaging: an archive file
.I package-x.y.z.tar.gz
unbundles into a directory named
.I package-x.y.z
in which there is an optional
.BR configure (1)
script, and a
.IR Makefile .
Running
.BR make (1)
in that directory should build the package, and
.B "make check"
(overridable with the
.BI \-\-check " target(s)"
option) should run validation tests.
.PP
.B build-all
recognizes archives in these formats:
.IR .jar ,
.IR .tar ,
.IR .tar.bz2 ,
.IR .tar.gz ,
.IR .tar.Z ,
and
.IR .zip .
It is the user's responsibility to ensure with suitable
shell startup files that on each remote host, GNU
.BR tar (1)
and
.BR unzip (1)
are found first in the search path, since they are needed to
unbundle packages.
.PP
Installation after builds is
.I not
automated; it is imperative to examine the build logs before
deciding whether it is safe to issue
.B "make install"
in a remote build directory.
.\" ========================================================
.SH "FILES"
In this section,
.I $REMOTEHOME
refers to the home directory on a remote build host.
.PP
Comments are supported in all of these files: they run from
sharp (#) to end of line.
.PP
.\" --------------------------------------------------------
.TP  \w'\fI$REMOTEHOME/.build/begin\fP'u+1n
.I $REMOTEHOME/.build/begin
Commands to be executed in the context of the login shell on
the remote host before the build begins.  This file can be
used to augment the search path,
.IR PATH ,
for shells like
.BR ksh (1)
and
.BR sh (1)
that do not provide for startup files for other than login
sessions.  Any output produced will be recorded near the
start of the build log file.
.\" --------------------------------------------------------
.TP
.I $HOME/.build/directories
The list of local directories to search for package archive
files.  If that file cannot be found, a default list of
source directories is provided.
.\" --------------------------------------------------------
.TP
.I $REMOTEHOME/.build/end
Commands to be executed in the context of the login shell on
the remote host after the build ends.  Any output produced
will be recorded near the end of the build log file.
.\" --------------------------------------------------------
.TP
.I $HOME/.build/userhosts
The default list of build hosts, with entries like this:
.IP
.RS
.RS
.nf
\&\fCjones@example.com:/local/build,$HOME/.build/c99\fP
.fi
.RE
.RE
.IP
The username, build directory, and environment file are
optional.  Except for the remote build directory, all files
listed are on the initiating host.
.\" --------------------------------------------------------
.TP
.I envfile
Environment files contain settings of environment variables
for
.BR configure (1)
on the remote systems.  Filenames are arbitrary: they are
set in entries in the
.I $HOME/.build/userhosts
file or similar files specified by the
.B \-\-userhosts
option, or else in command-line
.B \-\-on
options. For example, an environment file with
.IP
.RS
.RS
.nf
CC=c99
CFLAGS="-O3 -D_POSIX_SOURCE"
CXX=CC
CXXFLAGS="-O3 -D_POSIX_SOURCE"
.fi
.RE
.RE
.IP
specifies the C and C++ compilers, and compiler optimization
flags, to be used for the build.
.\" ========================================================
.SH "EXAMPLES"
Build two packages everywhere:
.RS
.nf
\&\fCbuild-all coreutils-5.0.0 gawk-3.1.2\fP
.fi
.RE
Build package on a specific host:
.RS
.nf
\&\fCbuild-all --on loaner.example.com gnupg-1.2.1\fP
.fi
.RE
Build package from nonstandard location:
.RS
.nf
\&\fCbuild-all --source $HOME/work butter-0.3.7\fP
.fi
.RE
Build package in POSIX compilation environment:
.RS
.nf
\&\fCbuild-all --userhosts posixhosts numlib-3.2.1\fP
.fi
.RE
.\" ========================================================
.SH "SEE ALSO"
.BR autoconf (1),
.BR autoheader (1),
.BR automake (1),
.BR configure (1),
.BR jar (1),
.BR ksh (1),
.BR make (1),
.BR scp (1),
.BR sh (1),
.BR ssh (1),
.BR tar (1),
.BR unzip (1).
.\" ========================================================


Original/sh/build-all.sh
#! /bin/sh -
# Build one or more packages in parallel on one or more build hosts.
#
# Usage:
#	build-all [ --? ]
#		  [ --all "..." ]
#		  [ --cd "..." ]
#		  [ --check "..." ]
#		  [ --configure "..." ]
#		  [ --environment "..." ]
#		  [ --help ]
#		  [ --logdirectory dir ]
#		  [ --on "[user@]host[:dir][,envfile] ..." ]
#		  [ --source "dir ..." ]
#		  [ --userhosts "file(s)" ]
#		  [ --version ]
#		  package(s)
#
# Optional initialization files:
#	$HOME/.build/directories	list of source directories
#	$HOME/.build/userhosts		list of [user@]host[:dir][,envfile]

IFS='
 	'

PATH=/usr/local/bin:/bin:/usr/bin
export PATH

UMASK=002
umask $UMASK

build_one()
{
	# Usage:
	#	build_one [user@]host[:build-directory][,envfile]

	arg="`eval echo $1`"

	userhost="`echo $arg | sed -e 's/:.*$//'`"

	user="`echo $userhost | sed -e s'/@.*$//'`"
	test "$user" = "$userhost" && user=$USER

	host="`echo $userhost | sed -e s'/^[^@]*@//'`"

	envfile="`echo $arg | sed -e 's/^[^,]*,//'`"
	test "$envfile" = "$arg" && envfile=/dev/null

	builddir="`echo $arg | sed -e s'/^.*://' -e 's/,.*//'`"
	test "$builddir" = "$arg" && builddir=/tmp

	parbase=`basename $PARFILE`

	# NB: update find_package() if these patterns are changed
	package="`echo $parbase | \
		sed	-e 's/[.]jar$//' \
			-e 's/[.]tar[.]bz2$//' \
			-e 's/[.]tar[.]gz$//' \
			-e 's/[.]tar[.]Z$//' \
			-e 's/[.]tar$//' \
			-e 's/[.]tgz$//' \
			-e 's/[.]zip$//'`"

	## echo "DEBUG: LOGDIR   = $LOGDIR"
	## echo "DEBUG: PARFILE  = $PARFILE"
	## echo "DEBUG: PAR      = $PAR"
	## echo "DEBUG: package  = $package"
	## echo "DEBUG: arg      = $arg"
	## echo "DEBUG: userhost = $userhost"
	## echo "DEBUG: host     = $host"
	## echo "DEBUG: builddir = $builddir"

	# Copy the package file if we cannot see it on the remote host
	echo $SSH $SSHFLAGS $userhost "test -f $PARFILE"
	if $SSH $SSHFLAGS $userhost "test -f $PARFILE"
	then
		parbaselocal=$PARFILE
	else
		parbaselocal=$parbase
		echo $SCP $PARFILE $userhost:$builddir
		$SCP $PARFILE $userhost:$builddir
	fi

	# Unbundle the archive file on the remote host, build, and
	# check it, running in the background

	sleep 1			# to guarantee unique log filename
	now="`date $DATEFLAGS`"
	logfile="$package.$host.$now.log"

	## echo "DEBUG: LOGFILE  = $LOGDIR/$package.$host.$now.log"

	nice $SSH $SSHFLAGS $userhost "
			echo '============================================================' ;
		        test -f $BUILDBEGIN && . $BUILDBEGIN || test -f $BUILDBEGIN && source $BUILDBEGIN || true ;
			echo 'Package:                $package' ;
			echo 'Archive:                $PARFILE' ;
			echo 'Date:                   $now' ;
			echo 'Local user:             $USER' ;
			echo 'Local host:             `hostname`' ;
			echo 'Local log directory:    $LOGDIR' ;
			echo 'Local log file:         $logfile' ;
			echo 'Remote user:            $user' ;
			echo 'Remote host:            $host' ;
			echo 'Remote directory:       $builddir' ;
			printf 'Remote date:            ' ;
			date $DATEFLAGS ;
			printf 'Remote uname:           ' ;
			uname -a || true ;
			printf 'Remote gcc version:     ' ;
			gcc --version | head -n 1 || echo ;
			printf 'Remote g++ version:     ' ;
			g++ --version | head -n 1 || echo ;
			echo 'Configure environment:  `$STRIPCOMMENTS $envfile | $JOINLINES`' ;
			echo 'Extra environment:      $EXTRAENVIRONMENT' ;
			echo 'Configure directory:    $CONFIGUREDIR' ;
			echo 'Configure flags:        $CONFIGUREFLAGS' ;
			echo 'Make all targets:       $ALLTARGETS' ;
			echo 'Make check targets:     $CHECKTARGETS' ;
			echo 'Disk free report for $builddir/$package:' ;
			df $builddir | $INDENT ;
			echo 'Environment:' ;
			env | env LC_ALL=C sort | $INDENT ;
			echo '============================================================' ;
			umask $UMASK ;
			cd $builddir || exit 1 ;
			/bin/rm -rf $builddir/$package ;
			$PAR $parbaselocal ;
			test "$parbase" = "$parbaselocal" && /bin/rm -f $parbase ;
			cd $package/$CONFIGUREDIR || exit 1 ;
			test -f configure && \
				chmod a+x configure && \
					env `$STRIPCOMMENTS $envfile | $JOINLINES` \
						$EXTRAENVIRONMENT \
							nice time ./configure $CONFIGUREFLAGS ;
			nice time make $ALLTARGETS && nice time make $CHECKTARGETS ;
			echo '============================================================' ;
			echo 'Disk free report for $builddir/$package:' ;
			df $builddir | $INDENT ;
			printf 'Remote date:            ' ;
			date $DATEFLAGS ;
			cd ;
		        test -f $BUILDEND && . $BUILDEND || test -f $BUILDEND && source $BUILDEND || true ;
			echo '============================================================' ;
		" < /dev/null > "$LOGDIR/$logfile" 2>&1 &
}

error()
{
	echo "$@" 1>&2
	usage_and_exit 1
}

find_file()
{
	# Usage:
	#	find_file file program-and-args
	# Return 0 (success) if found, 1 (failure) if not found

	if test -r "$1"
	then
		PAR="$2"
		PARFILE="$1"
		return 0
	else
		return 1
	fi
}

find_package()
{
	# Usage: find_package package-x.y.z
	base=`echo "$1" | sed -e 's/[-_][.]*[0-9].*$//'`
## echo "DEBUG: find_package $1"
## echo "DEBUG: SRCDIRS = $SRCDIRS"
	PAR=
	PARFILE=
	for srcdir in $SRCDIRS
	do
		test "$srcdir" = "." && srcdir="`pwd`"
## echo "DEBUG: srcdir = $srcdir"

		for subdir in "$base" ""
		do
			# NB: update package setting in build_one() if this list changes
			find_file $srcdir/$subdir/$1.tar.gz	"tar xfz"	&& return
			find_file $srcdir/$subdir/$1.tar.Z	"tar xfz"	&& return
			find_file $srcdir/$subdir/$1.tar	"tar xf"	&& return
			find_file $srcdir/$subdir/$1.tar.bz2	"tar xfj"	&& return
			find_file $srcdir/$subdir/$1.tgz	"tar xfz"	&& return
			find_file $srcdir/$subdir/$1.zip	"unzip -q"	&& return
			find_file $srcdir/$subdir/$1.jar	"jar xf"	&& return
		done
	done
}

set_userhosts()
{
	# Usage: set_userhosts file(s)
	for u in "$@"
	do
		if test -r "$u"
		then
			ALTUSERHOSTS="$ALTUSERHOSTS $u"
		elif test -r "$BUILDHOME/$u"
		then
			ALTUSERHOSTS="$ALTUSERHOSTS $BUILDHOME/$u"
		else
			error "File not found: $u"
		fi
	done
}

usage()
{
	cat <<EOF
Usage:
	$PROGRAM [ --? ]
			[ --all "..." ]
			[ --cd "..." ]
			[ --check "..." ]
			[ --configure "..." ]
			[ --environment "..." ]
			[ --help ]
			[ --logdirectory dir ]
			[ --on "[user@]host[:dir][,envfile] ..." ]
			[ --source "dir ..." ]
			[ --userhosts "file(s)" ]
			[ --version ]
			package(s)
EOF
}

usage_and_exit()
{
	usage
	exit $1
}

version()
{
	echo "$PROGRAM version $VERSION"
}

warning()
{
	echo "$@" 1>&2
	EXITCODE=`expr $EXITCODE + 1`
}

ALLTARGETS=
altlogdir=
altsrcdirs=
ALTUSERHOSTS=
BUILDBEGIN=./.build/begin
BUILDEND=./.build/end
BUILDHOME=$HOME/.build
CHECKTARGETS=check
CONFIGUREDIR=.
CONFIGUREFLAGS=
DATEFLAGS="+%Y.%m.%d.%H.%M.%S"
EXITCODE=0
EXTRAENVIRONMENT=
INDENT="awk '{ print \"\t\t\t\" \$0 }'"
JOINLINES="tr '\n' '\040'"
LOGDIR=
PROGRAM=`basename $0`
SCP=scp
SSH=ssh
# SCP="echo scp"
# SSH="echo ssh"
SSHFLAGS=${SSHFLAGS--x}
STRIPCOMMENTS='sed -e s/#.*$//'
userhosts=
VERSION=1.0

# Default initialization files
defaultdirectories=$BUILDHOME/directories
defaultuserhosts=$BUILDHOME/userhosts

# List of places to find package distributions, with a default
# list if the user has no personalized list:
SRCDIRS="`$STRIPCOMMENTS $defaultdirectories 2> /dev/null`"
test -z "$SRCDIRS" && \
	SRCDIRS="
			.
			/usr/local/src
			/usr/local/gnu/src
			$HOME/src
			$HOME/gnu/src
			/tmp
			/usr/tmp
			/var/tmp
		"

while test $# -gt 0
do
	case $1 in

	--all | --al | --a | -all | -al | -a )
		shift
		ALLTARGETS="$1"
		;;

	--cd | -cd )
		shift
		CONFIGUREDIR="$1"
		;;

	--check | --chec | --che | --ch | -check | -chec | -che | -ch )
		shift
		CHECKTARGETS="$1"
		;;

	--configure | --configur | --configu | --config | --confi | \
	--conf | --con | --co | \
	-configure | -configur | -configu | -config | -confi | \
	-conf | -con | -co )
		shift
		CONFIGUREFLAGS="$1"
		;;

	--environment | --environmen | --environme | --environm | --environ | \
	--enviro | --envir | --envi | --env | --en | --e | \
	-environment | -environmen | -environme | -environm | -environ | \
	-enviro | -envir | -envi | -env | -en | -e )
		shift
		EXTRAENVIRONMENT="$1"
		;;

	--help | --hel | --he | --h | '--?' | -help | -hel | -he | -h | '-?' )
		usage_and_exit 0
		;;

	--logdirectory | --logdirector | --logdirecto | --logdirect | \
	--logdirec | --logdire | --logdir | --logdi | --logd | --log | \
	--lo | --l | \
	-logdirectory | -logdirector | -logdirecto | -logdirect | \
	-logdirec | -logdire | -logdir | -logdi | -logd | -log | -lo | -l )
		shift
		altlogdir="$1"
		;;

	--on | --o | -on | -o )
		shift
		userhosts="$userhosts $1"
		;;

	--source | --sourc | --sour | --sou | --so | --s | \
	-source | -sourc | -sour | -sou | -so | -s )
		shift
		altsrcdirs="$altsrcdirs $1"
		;;

	--userhosts | --userhost | --userhos | --userho | --userh | \
	--user | --use | --us | --u | \
	-userhosts | -userhost | -userhos | -userho | -userh | \
	-user | -use | -us | -u )
		shift
		set_userhosts $1
		;;

	--version | --versio | --versi | --vers | --ver | --ve | --v | \
	-version | -versio | -versi | -vers | -ver | -ve | -v )
		version
		exit 0
		;;

	-*)
		error "Unrecognized option: $1"
		;;

	*)
		break
		;;

	esac
	shift
done

# Find a suitable mail client
for MAIL in /bin/mailx /usr/bin/mailx /usr/sbin/mailx /usr/ucb/mailx /bin/mail /usr/bin/mail
do
	test -x $MAIL && break
done
test -x $MAIL || error "Cannot find mail client"

# Command-line source directories precede defaults
SRCDIRS="$altsrcdirs $SRCDIRS"

if test -n "$userhosts"
then
	test -n "$ALTUSERHOSTS" &&
	    userhosts="$userhosts `$STRIPCOMMENTS $ALTUSERHOSTS 2> /dev/null`"
else
	test -z "$ALTUSERHOSTS" && ALTUSERHOSTS="$defaultuserhosts"
	userhosts="`$STRIPCOMMENTS $ALTUSERHOSTS 2> /dev/null`"
fi

# Check for something to do
test -z "$userhosts" && usage_and_exit 1

for p in "$@"
do
	find_package "$p"

	if test -z "$PARFILE"
	then
		warning "Cannot find package file $p"
		continue
	fi

	LOGDIR="$altlogdir"
	if test -z "$LOGDIR" -o ! -d "$LOGDIR" -o ! -w "$LOGDIR"
	then
		for LOGDIR in "`dirname $PARFILE`/logs/$p" $BUILDHOME/logs/$p /usr/tmp /var/tmp /tmp
		do
			test -d "$LOGDIR" || mkdir -p "$LOGDIR" 2> /dev/null
			test -d "$LOGDIR" -a -w "$LOGDIR" && break
		done
	fi

	msg="Check build logs for $p in `hostname`:$LOGDIR"
	echo "$msg"
	echo "$msg" | $MAIL -s "$msg" $USER 2> /dev/null

	for u in $userhosts
	do
		build_one $u
	done
done

# Limit exit status to common Unix practice
test $EXITCODE -gt 125 && EXITCODE=125

exit $EXITCODE
